### PR TITLE
TransposeLDS 

### PR DESCRIPTION
**Pull request by ramjana**
- TransposeLDS support for TLU=0

### CODE REVIEW

File: Tensile/Common.py [hunk 1]
```diff
@@ -703,6 +703,14 @@
...
     "LdsPadB":                     [ -1, 0, 1, 2, 3, 4, 8],
 
+    # Padding boundary for LDS. defines block-size for pad insertion. for every 'LdsBlockSizePerPad' bytes, LDS padding (pad value from LdsPad parameter)
+    # is added (readOffset aware of the pad and adjusts offset value based on this parameter value).good rule of thumb is LdsBlockSizePerPad >= unrollDepth * BPE
+    # optimized value is 128
+    "LdsBlockSizePerPad":          [-1, 64, 128, 256],
+
+    #Transpose LDS format. format to to store LDS in same memory format as global memory format. applicable only in TLU=0 case
+    "TransposeLDS":                [-1, 1, 0],
+
```
Comment (bensander) [inline Tensile/Common.py]: How do we add no padding - should this allow a value of 0?
How does this interact with existing LdsPadA/B, which IIRC are pads specified in terms of elements

File: Tensile/Code.py [hunk 2]
```diff
@@ -281,6 +281,61 @@ class LocalReadInst (Inst):
...
+  def  __init__(self,kernel,aIdx,bIdx,PLRval,innerUnroll):
+       self.endLine = ""
+       self.version = globalParameters["CurrentISA"]
+       self.kernel  = kernel
+       self.aIdx    = aIdx
+       self.bIdx    = bIdx
+       self.PLR     = PLRval
+       self.innerUnroll = innerUnroll
+
+  #def toCodeInst(self,kernel,aIdx,bIdx,PLRval,innerUnroll):
```
Comment (bensander) [inline Tensile/Code.py]: remove

File: Tensile/Code.py [hunk 3]
```diff
@@ -281,6 +281,61 @@ class LocalReadInst (Inst):
...
+
+  """
+  def  __init__(self,kernel,aIdx,bIdx,PLRval,innerUnroll):
+       self.endLine = ""
+       self.version = globalParameters["CurrentISA"]
+       self.kernel  = kernel
+       self.aIdx    = aIdx
+       self.bIdx    = bIdx
+       self.PLR     = PLRval
+       self.innerUnroll = innerUnroll
```
Comment (bensander) [inline Tensile/Code.py]: should this call the Inst __init__ constructor?

File: Tensile/KernelWriterSource.py [hunk 4]
```diff
@@ -2411,7 +2411,7 @@ def localReadInc(self, kernel, iui, tP):
   ##############################################################################
   # Local Read: Do It A/B
   ##############################################################################
-  def localReadDo(self, kernel, black, iui, epsi, tP):
+  def localReadDo(self, kernel, black, iui, epsi, uIdx, tP):
```
Comment (bensander) [inline Tensile/KernelWriterSource.py]: document uIdx in header

File: Tensile/SolutionStructs.py [hunk 5]
```diff
@@ -2485,11 +2489,27 @@ def assignDerivedParameters(state):
       state["LdsPadB"] = 0 if state["ProblemType"]["TLUB"] else state["VectorWidth"]
       assert(state["LdsPadB"] >= 0)
 
+    if "MatrixInstM" in state:
```
Comment (bensander) [inline Tensile/SolutionStructs.py]: Should this check for MatrixInst equal to non-zero value?  This seem a too-subtle way to check for MatrixInst enabled

File: Tensile/SolutionStructs.py [hunk 6]
```diff
@@ -1814,9 +1814,13 @@ def setGlobalLoadTileDimClassic(state, tc, numLoads, totalVectorsCoalesced, tota
           // state["NumLoadsCoalesced%s"%tc]
       state["LSP%s"%tc] = int(math.ceil(float(state["DepthU"]) / state["NumLoadsPerpendicular%s"%tc]))
     else:
-      state["LSC%s"%tc] = int(math.ceil(float(state["DepthU"]) / state["NumLoadsCoalesced%s"%tc]))
-      state["LSP%s"%tc] = state["MacroTile%s"%tc] \
-          // state["NumLoadsPerpendicular%s"%tc]
+      if state["TransposeLDS"] == 1:
+        state["LSC%s"%tc] = int(math.ceil(float(state["DepthU"]) / state["NumLoadsCoalesced%s"%tc]))
+        state["LSP%s"%tc] = min(state["MacroTile%s"%tc], ((state["NumThreads"]*state["GlobalReadVectorWidth"])//state["DepthU"]))
```
Comment (bensander) [inline Tensile/SolutionStructs.py]: seems transpose would need to change both LSC and LSP ?

File: Tensile/Common.py (hunk 1 above)
Comment (ramjana) [reply Tensile/Common.py]: transposeLDS use LdsPad parameter to determine padding size (ldsPad*BPE) to be added to lwo. transposeLDS=0 branches the code old implementation, i.e swizzle (write perp elements) global fetch elements during local write

File: Tensile/Code.py (hunk 3 above)
Comment (ramjana) [reply Tensile/Code.py]: will do

File: Tensile/Code.py (hunk 2 above)
Comment (ramjana) [reply Tensile/Code.py]: done

File: Tensile/SolutionStructs.py (hunk 5 above)
Comment (ramjana) [reply Tensile/SolutionStructs.py]: we are using this approach many places. we should have derived parameter like MatrixInstruction for checks.

File: Tensile/KernelWriterSource.py (hunk 4 above)
Comment (ramjana) [reply Tensile/KernelWriterSource.py]: done..

File: Tensile/SolutionStructs.py (hunk 6 above)
Comment (ramjana) [reply Tensile/SolutionStructs.py]: fixed.. do not need to change for TRansposeLDS=1

File: Tensile/Code.py [hunk 7]
```diff
@@ -281,6 +281,52 @@ class LocalReadInst (Inst):
...
+
+  def __str__(self):
+      # single precision
+      kStr = ""
+      numOfRowsperMfma = 1
+      numOfColsperMfma = kernel["MatrixInstN"]
+      numOfRowInsts = kernel["ThreadTile0"]/numberOfRowsperMfma
+      numOfColInsts = kernel["ThreadTile1"]/numberOfColsperMfma
+      numOfDstRgs = (kernel["MatrixInstN"] * kernel["MatrixInstM"] * kernel["MatrixInstB"] // globalParameters["WavefrontWidth"])
+      if self.kernel["ProblemType"]["DataType"].isSingle():
```
Comment (AlexBrownAMD) [inline Tensile/Code.py]: pyflakes is complaining about this stuff

File: Tensile/KernelWriterAssembly.py [hunk 8]
```diff
@@ -6819,8 +7105,13 @@ def calculateLdsWriteOffset(self, perp, para, sPerp, sPara, kernel, tP, localWri
       lspaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
       lspaOffset += rem * ldl + perp_rem
     else:
-      lscaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
-      lscaOffset += rem
+      if not kernel["TransposeLDS"]:
+        lscaOffset *= (kernel[tP["mt"]] + kernel["LdsPad%s"%tc])
+        lscaOffset += rem
+      else:
+ 	#FIXME re-verify lscaOffset=0
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py]: Is this done?

Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: may need to revisit with extensive testing..

File: Tensile/Code.py (hunk 3 above)
Comment (ramjana) [reply Tensile/Code.py]: done

File: Tensile/KernelWriterAssembly.py [hunk 9]
```diff
@@ -3979,6 +4011,29 @@ def graFinalOffsets(self, kernel, tP):
...
+          else:
+            # Fractional load use the more accurate lsc, multiply by VW later
+            divisorName = tP["lsc"]
+        else:
+          if tP["grcv"]:
+            divisorName = tP["lsp"]
+          else:
+            divisorName = tP["lvp"]
+        numPerpElementsPerLoad = kernel[divisorName]
+        numPerpElementsPerWave = tP["nrp"]*numPerpElementsPerLoad
```
Comment (jichangjichang) [inline Tensile/KernelWriterAssembly.py]: should be 
numPerpElementsPerWave = tP["nrp"] * globalParameters["WavefrontWidth"] // numPerpElementsPerLoad
kernel['LVCA'] mean how many thread load in unroll dim.
"globalParameters["WavefrontWidth"] // kernel['LVCA']" mean how many M lines can be loaded for 1 Wavefront in one buffer load instruction.
(tP["nrp"] * globalParameters["WavefrontWidth"] // numPerpElementsPerLoad) mean how many M lines can be loaded for 1 Wavefront.

File: Tensile/KernelWriterAssembly.py [hunk 10]
```diff
@@ -4760,28 +4888,150 @@ def lraFinalOffset(self, kernel, tP):
...
+            else:
+              # Fractional load use the more accurate lsc, multiply by VW later
+              divisorName = tP["lsc"]
+          else:
+            if tP["grcv"]:
+              divisorName = tP["lsp"]
+            else:
+              divisorName = tP["lvp"]
+          numPerpElementsPerLoad = kernel[divisorName]
+          numPerpElementsPerWave = tP["nrp"]*numPerpElementsPerLoad
```
Comment (jichangjichang) [inline Tensile/KernelWriterAssembly.py]: should be 
numPerpElementsPerWave = tP["nrp"]* globalParameters["WavefrontWidth"] // numPerpElementsPerLoad

File: Tensile/KernelWriterAssembly.py (hunk 9 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: had fix in directToLds , fix is to use LSP/LVP parameters to calculate number of PerpelementsPerWave

File: Tensile/KernelWriterAssembly.py (hunk 10 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: fixed.

**Pull request merged**